### PR TITLE
unstake / unbond from waiting

### PR DESCRIPTION
**Pull request by sasurobert**
Added new feature to unStake and unBond from additional queue without waiting the unBond period. The same goes for unStakeTokens and unBondTokens if there are no more nodes registered for the given validator.

### CODE REVIEW

File: vm/systemSmartContracts/validator.go [hunk 1]
```diff
@@ -1124,9 +1127,35 @@ func (v *validatorSC) unStakeNodesFromStakingSC(blsKeys [][]byte, registrationDa
...
+func (v *validatorSC) computeProcessUnStakeFromNodes(
+	registrationData *ValidatorDataV2,
+	validatorConfig ValidatorConfig,
+	numNodes uint64,
+	unStakeNonce uint64,
+) vmcommon.ReturnCode {
+	if numNodes == 0 {
+		return vmcommon.Ok
+	}
+	unStakeFromActiveNodes := big.NewInt(0).Mul(validatorConfig.NodePrice, big.NewInt(0).SetUint64(numNodes))
```
Comment (SebastianMarian) [inline vm/systemSmartContracts/validator.go:1153]: unStakeFromNodes as you call this method either with numSuccessFromWaiting or with numSuccessFromActive

File: vm/systemSmartContracts/validator.go [hunk 2]
```diff
@@ -1124,9 +1127,35 @@ func (v *validatorSC) unStakeNodesFromStakingSC(blsKeys [][]byte, registrationDa
...
+		if stakedData.UnStakedNonce > currentNonce {
+			numSuccessFromWaiting++
+		}
+	}
+
+	numSuccessFromActive := numSuccess - numSuccessFromWaiting
+	return numSuccessFromActive, numSuccessFromWaiting
+}
+
+func (v *validatorSC) computeProcessUnStakeFromNodes(
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1144]: maybe rename as computeProcess... doesn't tell much :D

File: vm/systemSmartContracts/validator.go [hunk 3]
```diff
@@ -1124,9 +1127,35 @@ func (v *validatorSC) unStakeNodesFromStakingSC(blsKeys [][]byte, registrationDa
 		}
 
 		numSuccess++
+
+		stakedData, err := v.getStakedData(blsKey)
+		if err != nil {
+			continue
+		}
+		if stakedData.UnStakedNonce > currentNonce {
```
Comment (iulianpascalau) [inline vm/systemSmartContracts/validator.go:1135]: As found out by @AdoAdoAdo & me this is always 0 if the bls key was never staked. UnStakedEpoch will hold core.DefaultUnstakedEpoch

Comment (sasurobert) [reply vm/systemSmartContracts/validator.go]: fixed this.

**Review by iulianpascalau — APPROVED** [review]

File: vm/systemSmartContracts/validator.go (hunk 1 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1151]: done

File: vm/systemSmartContracts/validator.go (hunk 2 above)
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1142]: I do not have a better idea

Comment (sasurobert) [reply vm/systemSmartContracts/validator.go]: renamed

File: vm/systemSmartContracts/validator.go [hunk 4]
```diff
@@ -1105,8 +1105,11 @@ func (v *validatorSC) basicChecksForUnStakeNodes(args *vmcommon.ContractCallInpu
 	return registrationData, vmcommon.Ok
 }
 
-func (v *validatorSC) unStakeNodesFromStakingSC(blsKeys [][]byte, registrationData *ValidatorDataV2) uint64 {
+func (v *validatorSC) unStakeNodesFromStakingSC(blsKeys [][]byte, registrationData *ValidatorDataV2) (uint64, uint64) {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply vm/systemSmartContracts/validator.go:1108]: we do not really use private DTOs in the project. these are private local functions and someone could look what is inside.

**Review by iulianpascalau — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Review by stefanandy — APPROVED** [review]
System Test Passed

**Pull request merged**